### ISSUE TITLE
Listview items are only rebuilt if they are not in the viewport/leave the viewport

### ISSUE THREAD

**Issue by nicmohr25** [FlutterFlow/flutterflow-issues#1440]
### Has your issue been reported?

- [X] I have searched the existing issues and confirm it has not been reported.
- [X] I give permission for members of the FlutterFlow team to access and test my project for the sole purpose of investigating this issue.

### Current Behavior

I have in total four list views in one of my pages. Some Listviews can show the same document data (listitem). The items are generated from an API Response. Every Listitem is represented by a component
If i change a value and clear the Query Cache, the items are only rebuilt when they are outside of the current viewport or if they are scrolled out if the viewport and back in the viewport. 

### Expected Behavior

Items should also be rebuilt when they are inside the viewport

### Steps to Reproduce

1. Create Page
2. Add Two List Views
4. Generate children from API response
5. Change the value with a POST so when the API is triggered again it return another value. 
6. Value is not changing when items is in viewport
7. Play with the list and scroll back and forth
8. Item will be rebuild as soon as its outside of the viewport

### Reproducible from Blank

- [X] The steps to reproduce above start from a blank project.

### Bug Report Code (Required)

ITFTlcmB34hNodcF+KrqdcJslWQsCUMdXOczitVEew4bCIj4OrMuPPTBP1VtZ+KzaHx2EVaYikEB+MqOk/HpUPUEaAmdUbpa1KoIWBPPbDmVMozTEs2eS3dDL99bfRW60reFmxJSCLBeWE5h2Dn3N+eiTT+Cf9qOYwx5e6fDbOY=

### Context

Showing an icon depending on a bool value that is returned by the API response.

### Visual documentation

N/A

### Environment

```markdown
- FlutterFlow version: FlutterFlow v3.1 - Flutter 3.10.4
- Platform: Web
- Browser name and version: Chrome Version 116.0.5845.179 (Offizieller Build) (x86_64)
- Operating system and version affected: MacOS Ventura Version 12.4.1
```

**Comment by hariprasadms** [comment]
Hi @nicmohr25 - I appreciate you bringing this up. Api calls are future (asynchronous) calls, therefore when the data changes, the list-widget must be rebuilt in order for the updated data to display. Either you should have an action to rebuild the list and update its status, or you should navigate away from the page and back to it so you can see the change. I can confirm that the List-View is working as expected. This is not a bug.

**Comment by nicmohr25** [comment]
Hello @hariprasadms, thanks for your answer. 
But shouldn't the ListView rebuild its items when the api call is refreshed? 
In the Network Tab i can see that the API call is resend and the correct values are returned by my endpoint. 
But the Component still shows the wrong state. When i move the corresponding item outside of the Viewport, that particular item is rebuild and is displaying the correct values.

**Comment by nicmohr25** [comment]
If anyone should ever have a similar problem: 
If you use components within a ListView, make sure to set the unique key manually. If the status within the component changes and this is updated to a backend, the API call is refreshed and the same status update should also be displayed in a second ListView, the ListView must be rendered. However, this only happens for components outside the viewport because they are not yet layed out and are going to be rendered as soon as they approach the Viewport. 
Apparently, the ListView checks the key of the components first. If the key did not change significantly, items within the viewport are not rendered.

Maybe that helps someone. 
 Cheers

**Comment by agreaves** [comment]
Hi @nicmohr25, thanks for the additional info. I can confirm this is because the unique key is going to prevent the widget from being rebuilt unless it's value changes with the new data. This is a challenging technical concept to convey so I appreciate you explaining it. Closing this as it's not a bug strictly speaking but we hope to add more education around this so folks are not confused at the behavior. Thanks again!

**Comment by jerperdue** [comment]
@agreaves @hariprasadms @nicmohr25 

I don't think I understand. Can you provide clarity on my scenario? I have a page with a listview and a details page. 

The listview has a component:
![Screenshot 2023-09-26 at 2 07 22 PM](https://github.com/FlutterFlow/flutterflow-issues/assets/62181919/230f98d7-fc3d-4025-976b-1a967e1b5c49)

The list item has a dropdown:
![Screenshot 2023-09-26 at 2 09 38 PM](https://github.com/FlutterFlow/flutterflow-issues/assets/62181919/4b489afd-48f8-49e3-ad22-5962f7695516)

If I navigate to the details page for the list item, change the status dropdown there, and hit "Back", the listview dropdown does not update with the new value (although I verified the value is updated on the backend). 

What is the solution here? Are you saying I simply need to put something as a Unique Key? Should it be a static value or something dynamic? What do you recommend doing there?
![Screenshot 2023-09-26 at 2 06 08 PM](https://github.com/FlutterFlow/flutterflow-issues/assets/62181919/5f5202f4-6c19-4a52-bafb-5837da674372)

**Comment by nicmohr25** [comment]
the ListView is not registering your change because for the ListView your Component has not changed.
The Unique Key is preventing that. What I did to solve this issue is to integrate the value which can change into my unique key.
Flutterflow builds the Unique Key by default like this: key: Key('Keyftn_${apiPostsIndex}of${apiPosts.length}')
What I did was to integrate the value that is critically into the unique key: Keyftn_${apiPostsIndex}of${apiPosts.length}_${your_critical_value}'.
With that, the ListView was able to register the change and rebuild the ListItem.
I don't think that it necessarily refreshes the API Request, but thats something you might need to check. This was just something I needed in my case.
